sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/StreamingProducer.cs

```diff
@@ -0,0 +1,377 @@
+﻿  // Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Azure.Messaging.EventHubs.Producer
+{
+    /// <summary>
+    /// This class allows the application to create and use a Streaming
+    /// producer instance.
+    /// </summary>
+    public class StreamingProducer : IAsyncDisposable
+    {
+        private EventHubProducerClient Producer;
+        private ConcurrentDictionary<string, ConcurrentQueue<EventData>> PartitionQueues;
+        private ConcurrentDictionary<string, EventDataBatch> PendingBatches;
+        private ConcurrentQueue<EventData> GeneralQueue;
+        private EventDataBatch GeneralPendingBatch;
+
+        /// <summary>
+        /// A handler for event batches that have been sent successfully.
+        /// </summary>
+        public event Action<PublishSucceededEventArgs> PublishSucceeded;
+
+        /// <summary>
+        /// A handler for event batches that have failed to successfully send after applying the
+        /// retry policy.
+        /// </summary>
+        public event Action<PublishFailedEventArgs> PublishFailed;
+
+        /// <summary>
+        /// The fully qualified Event Hubs namespace that this producer is currently associated with,
+        /// which will likely be similar to <c>{yournamespace}.servicebus.windows.net</c>.
+        /// </summary>
+        public string FullyQualifiedNamespace => Producer.FullyQualifiedNamespace;
+
+        /// <summary>
+        /// The name of the Event Hub that this producer is connected to, specific to the Event Hubs
+        /// namespace that contains it.
+        /// </summary>
+        public string EventHubName => Producer.EventHubName;
+
+        /// <summary>
+        /// </summary>
+        //public string Identifier => Producer.Identifier;
+        public string Identifier { get; protected set; }
+
+        /// <summary>
+        /// Returns the total number of events that are currently in the queue waiting to be sent.
+        /// queue waiting to be sent.
+        /// </summary>
+        public int TotalQueuedEventCount { get; protected set; }
+
+        /// <summary>
+        /// <c>true</c> if the streaming producer has been closed <c>false</c> otherwise.
+        /// </summary>
+        public bool IsClosed => Producer.IsClosed; //??
```

Comment:
We'll need to allow this to be overridden for unit testing purposes.
```suggestion
        public virtual int TotalQueuedEventCount { get; protected set; }

        /// <summary>
        /// <c>true</c> if the streaming producer has been closed <c>false</c> otherwise.
        /// </summary>
        public virtual bool IsClosed => Producer.IsClosed; //??
```